File: 449.序列化和反序列化二叉搜索树/main.go
package leetcode

import (
	"strconv"
	"strings"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Codec struct{}

func Constructor() Codec {
	return Codec{}
}

func (this *Codec) serialize(root *TreeNode) string {
	list := []string{}

	var iter func(*TreeNode, *[]string)
	iter = func(_node *TreeNode, _list *[]string) {
		if _node == nil {
			return
		}

		iter(_node.Left, _list)
		iter(_node.Right, _list)
		*_list = append(*_list, strconv.Itoa(_node.Val))
	}

	iter(root, &list)

	return strings.Join(list, "-")
}

func (this *Codec) deserialize(data string) *TreeNode {
    if data == "" {
        return nil
    }

    list := strings.Split(data, "-")

    var construct func(int, int) *TreeNode
    construct = func(lower, upper int) *TreeNode {
        if len(list) == 0 {
            return nil
        }

        val, _ := strconv.Atoi(list[len(list)-1])

        if val < lower || val > upper {
            return nil
        }

        list = list[:len(list)-1]
		
        return &TreeNode{Val: val, Right: construct(val, upper), Left: construct(lower, val)}
    }

    return construct(math.MinInt32, math.MaxInt32)
}

File: 449.序列化和反序列化二叉搜索树/main.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function serialize(root: TreeNode | null): string {
    const list = [];

    const iter = (_root: TreeNode | null, _list) => {
        if (!_root) return

        iter(_root.left, _list);
        iter(_root.right, _list);
        _list.push(_root.val);
    }

    iter(root, list);
    return list.join('-');
};

function deserialize(data: string): TreeNode | null {
    if (!data.length) return null;

    let list = data.split('-').map(_ => parseInt(_));

    const construct = (lower: number, upper: number, _list: number[]) => {
        if (!_list.length || _list[_list.length - 1] < lower || _list[_list.length - 1] > upper) return null;

        const value = _list.pop();
        const root = new TreeNode(value);
        root.right = construct(value, upper, _list);
        root.left = construct(lower, value, _list);
        return root;
    }

    return construct(-Number.MAX_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, list);
};